pkg/kubelet/cm/cpumanager/cpu_assignment.go

```diff
@@ -519,6 +547,60 @@ func (a *cpuAccumulator) takeFullSockets() {
 	}
 }
 
+func (a *cpuAccumulator) takeFullUncore() {
+	for _, uncore := range a.freeUncoreCache() {
+		cpusInUncore := a.topo.CPUDetails.CPUsInUncoreCaches(uncore)
+		if !a.needsAtLeast(cpusInUncore.Size()) {
+			continue
+		}
+		klog.V(4).InfoS("takeFullUncore: claiming uncore", "uncore", uncore)
+		a.take(cpusInUncore)
+	}
+}
+
+func (a *cpuAccumulator) takePartialUncore(uncoreID int) {
```

Comment:
The way I read this method it should be called `takeRestOfUncore` right?

It seems to be an optimization that tries to prevent leaving unusably small amount of cpus lying around free. Am I right?

Reply:
@wongchar correct me if I'm wrong, but that's not quite what's it's trying to do, it will check to see if we have some available to the uncore cache group and take what it needs but not exactly taking the rest of the uncore.

Reply:
Well.. maybe I do not understand the logic properly, but the if on row 569 seems to be checking whether the needed cpu count == (equals) the free cpus count.

Granted, the freeCores/freeCPUs computation is based on numCoresNeeded, but the semantics of the whole sequence is pretty complicated.

Maybe a better comment is needed.

Reply:
I did a few tests and the logic of the PR seems to be able to claim partial uncores overall, it is the naming and logic is not obvious from the code.

Reply:
The code (as I understood and LGTM'd) is meant to allocat a partially-available uncore, with some cpus already busy, depending on both the free cpus on the uncore and the requested quantity. The name should IMO contain somehow "partial" or a synonim, because after this code runs, it legitimately leave a uncore group partially allocated.

Reply:
Agree with comments above.

`takePartialUncore` is used when the number of cpus needed is less than a free uncore cache size (or when the remainder of cpus needed after `takeFullUncore` is less than a free uncore cache size). This is a best-effort method to keep the cpu assignment limited to a single uncore cache. 

Passing directly to `takeFullCores` or `takeRemainingCPUs` immediately after `takeFullUncore` could result in cpu assignments distributing across multiple uncore caches when it could fit within a single uncore cache.

Added some comments to `takePartialUncore` and 'takeUncoreCache' to help outline the comments above. Please let me know if this does not clear it up :)